File: src/com/rs/game/world/task/WorldTasksManager.java
package com.rs.game.world.task;

import java.util.Collections;
import java.util.LinkedList;
import java.util.List;

/**
 * This class manages all world tasks
 */
public class WorldTasksManager {
	
	/**
	 * The list of tasks that are being processed
	 */
	private static final List<WorldTaskInformation> TASKS = Collections.synchronizedList(new LinkedList<WorldTaskInformation>());
	
	/**
	 * Processes all the tasks in the world
	 */
	public static void processTasks() {
		for (WorldTaskInformation taskInformation : TASKS.toArray(new WorldTaskInformation[TASKS.size()])) {
			try {
				if (taskInformation.getInitialTickDelay() > 0) {
					taskInformation.setInitialTickDelay(taskInformation.getInitialTickDelay() - 1);
					continue;
				}
				// so we can store the ticks passed in the info class
				taskInformation.getTask().setTicksPassed(taskInformation.getTask().getTicksPassed() + 1);
				taskInformation.getTask().run();
				if (taskInformation.getTask().needRemove) {
					TASKS.remove(taskInformation);
				} else {
					taskInformation.setInitialTickDelay(taskInformation.getRepeatTickDelay());
				}
			} catch (Exception e) {
				TASKS.remove(taskInformation);
				e.printStackTrace();
			}
		}
	}
	
	/**
	 * Schedules a new task with a 0 initial delay and never repeating
	 */
	public static void schedule(WorldTask task) {
		schedule(task, 0, -1);
	}
	
	/**
	 * Schedules a task with a set delay count but never repeating
	 */
	public static void schedule(WorldTask task, int delayCount) {
		schedule(task, delayCount, -1);
	}
	
	/**
	 * Schedules a task with a set delay count and a set repeat count
	 */
	public static void schedule(WorldTask task, int delayCount, int periodCount) {
		if (task == null || delayCount < 0 || periodCount < 0) {
			return;
		}
		TASKS.add(new WorldTaskInformation(task, delayCount, periodCount));
	}
	
}

File: src/main/java/org/redrune/engine/tick/task/WorldTasksManager.kt
package org.redrune.engine.tick.task

import java.util.*

/**
 * This class manages all world tasks
 */
object WorldTasksManager {
    /**
     * The list of tasks that are being processed
     */
    private val TASKS = Collections.synchronizedList(LinkedList<WorldTaskInformation>())

    /**
     * Processes all the tasks in the world
     */
    fun processTasks() {
        for (taskInformation in TASKS.toTypedArray()) {
            try {
                if (taskInformation.initialTickDelay > 0) {
                    taskInformation.initialTickDelay = taskInformation.initialTickDelay - 1
                    continue
                }
                // so we can store the ticks passed in the info class
                taskInformation.task.ticksPassed = taskInformation.task.ticksPassed + 1
                taskInformation.task.run()
                if (taskInformation.task.isNeedRemove) {
                    TASKS.remove(taskInformation)
                } else {
                    taskInformation.initialTickDelay = taskInformation.repeatTickDelay
                }
            } catch (e: Exception) {
                TASKS.remove(taskInformation)
                e.printStackTrace()
            }
        }
    }
    /**
     * Schedules a task with a set delay count and a set repeat count
     */
    /**
     * Schedules a new task with a 0 initial delay and never repeating
     */
    /**
     * Schedules a task with a set delay count but never repeating
     */
    @JvmStatic
	@JvmOverloads
    fun schedule(task: WorldTask?, delayCount: Int = 0, periodCount: Int = -1) {
        if (task == null || delayCount < 0) {
            return
        }
        TASKS.add(WorldTaskInformation(task, delayCount, periodCount))
    }
}